modules/core/src/main/java/org/apache/ignite/cache/query/IndexQuery.java

```diff
@@ -0,0 +1,263 @@
+
+            cond.addCondition((String) vals[i], null, wrapNull(vals[i + 1]));
+        }
+
+        idxCond = cond;
+
+        return this;
+    }
+
+    /** Less Then or Equal. */
+    public IndexQuery<K, V> lte(String field, Object val) {
+        lt(field, val);
+
+        ((RangeIndexCondition) idxCond).upperInclusive(true);
+
+        return this;
+    }
+
+    /** Less Then or Equal. */
+    public IndexQuery<K, V> lte(String field, Object val, String field2, Object... vals) {
+        lt(field, val, field2, vals);
+
+        ((RangeIndexCondition) idxCond).upperInclusive(true);
+
+        return this;
+    }
+
+    /** Greater Then. */
+    public IndexQuery<K, V> gt(String field, Object val) {
+        A.ensure(idxCond == null, "The only index condition is supported.");
+
+        RangeIndexCondition cond = new RangeIndexCondition();
+
+        cond.addCondition(field, wrapNull(val), null);
+
+        idxCond = cond;
+
+        return this;
+    }
+
+    /** Greater Then. */
+    public IndexQuery<K, V> gt(String field, Object val, String field2, Object... vals) {
+        A.ensure(idxCond == null, "The only index condition is supported.");
+
+        A.notEmpty(vals, "vals");
+        A.ensure(vals.length % 2 == 1, "number of fields has to be equal number of values.");
+
+        RangeIndexCondition cond = new RangeIndexCondition();
+
+        cond.addCondition(field, wrapNull(val), null);
+        cond.addCondition(field2, wrapNull(vals[0]), null);
+
+        for (int i = 1; i + 1 < vals.length; i += 2) {
+            A.ensure(vals[i] instanceof String, "waited for field name but got " + vals[i]);
+
+            cond.addCondition((String) vals[i], wrapNull(vals[i + 1]), null);
+        }
+
+        idxCond = cond;
+
+        return this;
+    }
+
+    /** Greater Then or Equal. */
+    public IndexQuery<K, V> gte(String field, Object val) {
+        gt(field, val);
+
+        ((RangeIndexCondition) idxCond).lowerInclusive(true);
+
+        return this;
+    }
+
+    /** Greater Then or Equal. */
+    public IndexQuery<K, V> gte(String field, Object val, String field2, Object... vals) {
+        gt(field, val, field2, vals);
+
+        ((RangeIndexCondition) idxCond).lowerInclusive(true);
+
+        return this;
+    }
+
+    /** Between. Lower and upper boundaries are inclusive. */
+    public IndexQuery<K, V> between(String field, Object lower, Object upper) {
+        A.ensure(idxCond == null, "The only index condition is supported.");
+
+        RangeIndexCondition cond = new RangeIndexCondition();
+
+        cond.addCondition(field, wrapNull(lower), wrapNull(upper));
+
+        cond.lowerInclusive(true);
+        cond.upperInclusive(true);
+
+        idxCond = cond;
+
+        return this;
+    }
+
+    /** Between. Lower and upper boundaries are inclusive. */
+    public IndexQuery<K, V> between(String field, Object lower, Object upper, String field2, Object... vals) {
```

Comment:
Looks ugly for me. Can it be relpaced by several call of between with accumulate fields?

Reply:
I'm not sure that chaining is good approach. Code looks clearly but actually it's not true. For example:
```
IndexQuery.forIndex()
    .lt("fieldA", 1)
    .lt("fieldB", 2)
```

There has to be a complex validation that different chain parts should be consistent:
1. No gt() and lt() together on the same query;
2. From other side it's OK to combine between() with other range methods;
3. That max() or min() methods are the only methods.

Also Repeating multiple times "lt" declaration is weird. It's enough to declare condition once, then we need only provide list of fields. 

I think that alternative can be introducing a IndexField class that is argument of all index operations (varags too). 

```
IndexQuery().forIndex()
    .lt(field("fieldA", 1))
    .pageSize(1024);

IndexQuery().forIndex()
    .lt(
        field("fieldA", 1),
        field("fieldB", 2))
    .pageSize(1024);

IndexQuery().forIndex()
    .between(
        field("fieldA", 1, 2),
        field("fieldB", 2, 3))
    .pageSize(1024);

static IndexFieldCondition field(String field, Object val, Object? val2) { return new IndexFieldCondition(field, val, val2); }

```

In this case it a little bit verbose for single field index, but looks OK for multiple fields index. 

WDYT?

Reply:
@tledkov-gridgain  Implemented new API, that covers this case.